commit 0128cf0d79c6f5a3d17f32969df7cc81a31bb0ff (refs/changes/73/2273/1)
Author: Trent Mick <trentm@gmail.com>
Date:   2017-07-25T01:53:48+00:00 (2 years, 2 months ago)
    
    joyent/gerritbot#1 makecheckbot review is failing to post a review when there are make check results for files *not* in the patchset

diff --git a/.gitignore b/.gitignore
new file mode 100644
index 0000000..07e6e47
--- /dev/null
+++ b/.gitignore
@@ -0,0 +1 @@
+/node_modules
diff --git a/server.js b/server.js
index a75fe51..346d2d5 100644
--- a/server.js
+++ b/server.js
@@ -719,6 +719,47 @@ SlaveConnection.prototype.state_running.report = function (S) {
 		if (err) {
 			self.sc_log.error({ err: err },
 			    'failed to post review');
+			/*
+			 * If we hit
+			 * <https://bugs.chromium.org/p/gerrit/issues/detail?id=3475>
+			 * then fallback to a dumber report that doesn't try
+			 * to match 'make check' output to patchset files.
+			 * See arekinath/gerritbot#1.
+			 */
+			var marker = 'not found in revision';
+			if (err.message.search(marker) !== -1) {
+				S.gotoState('running.reportfallback');
+				return;
+			}
+		}
+		S.gotoState('closing');
+	}));
+};
+
+SlaveConnection.prototype.state_running.reportfallback = function (S) {
+	mod_assert.ok(this.sc_status !== 0,
+	    'this impl assumes the "make check" failed');
+
+	var self = this;
+	var review = {};
+	review.labels = {};
+	review.message = '"make check" exited with status ' + this.sc_status;
+	review.labels['CI-Testing'] = '-1';
+
+	var start = this.sc_out.length - 50;
+	if (start < 0)
+		start = 0;
+	var lines = this.sc_out.slice(start,
+	    this.sc_out.length);
+	lines = lines.map(function (v) { return (' ' + v); });
+	review.message += '\n\n' + lines.join('\n');
+
+	review.project = this.sc_change.project;
+	var spec = this.sc_change.number + ',' + this.sc_patchset.number;
+	gerrit.review(spec, review, S.callback(function (err) {
+		if (err) {
+			self.sc_log.error({ err: err },
+			    'failed to post review (reportfallback)');
 		}
 		S.gotoState('closing');
 	}));
